### ISSUE TITLE
Allow Skipping of Token Indexer

### ISSUE THREAD

**Issue by RXminuS** [allenai/allennlp#2498]
Currently a TokenIndexer must be provided to a TextField that returns integers (because they are forced to a Long Tensor). However, this behavior doesn't make sense for every text field. For instance, we've just made a TokenEmbedder that handles OOV tokens by doing a cool hashing trick. As such it's vocabulary is infinite and we just need the Token objects.

Currently, there is no nice way to handle this however other than making a special field type that doesn't always call the TokenIndexer. But then it becomes hard to compare "normal" models since you can't just swap out a Indexer & Embedder anymore and have to hard-code this behaviour in the Dataset reader.

We've tried several "hacky" solutions such as making a fake TokenIndexer that passes along the tokens but then we run into issues here https://github.com/allenai/allennlp/blob/89056f1e5063a3b10ab9230d8d6d0a43f5657fd0/allennlp/data/fields/text_field.py#L148 since a LongTensor is forced. There is a note for @matt-gardner already :-)

We need some help to discuss and potentially settle on a better design.

**Comment by RXminuS** [comment]
The idea that seems to work nicest for us ATM seems to be to allow a TokenIndexer to return a list of integers OR a torch.Tensor directly. Then we can implement our "embedder" as a Indexer instead which although it sounds weird initially actually makes a lot of sense...an embedding is in a way a dense index into a multi-dimensional space :-)

**Comment by matt-gardner** [comment]
I'm not totally sure what exactly you're trying to do.  We do handle hashing-type functionality in some of our code (see the way we handle byte encoding of characters [here](https://github.com/allenai/allennlp/blob/89056f1e5063a3b10ab9230d8d6d0a43f5657fd0/allennlp/data/tokenizers/character_tokenizer.py#L57-L60)), so there may be a way to get what you want without needing to change the API.  If you can give me some more detail on your use case, I can help better with what options there are.

**Comment by RXminuS** [comment]
Ok, in our specific case we use [Magnitude](https://github.com/plasticityai/magnitude/blob/master/ELMo.md) to produce torch.Tensors based on the text in a Token but also hashing POS tags etc. For OOV tokens it simply uses a hashing trick to produce a deterministic vector that is close to other vectors with similar character n-grams.

For this to work our TokenEmbedder needs to actually be passed Token objects rather than integer indexes; since the vocabulary is infinite. I guess you could theoretically index all the tokens as characters and use the offset value to later join them together again, but the problem is that there are all these other things (POS tags etc) in a Token object that we might want to also "Embed".

For now we made a small change to the `TextField` to do
```python
indexer_tensors = {key: tensor if issubclass(tensor, torch.Tensor) else torch.LongTensor(tensor) for key, tensor in padded_array.items()}
```
where
```python
self._indexed_tokens: Optional[Dict[str, Union[TokenList, torch.Tensor]]] = None
```

Our MagnitudeEmbedder is now a TokenIndexer who's `forward` is `(tokens: List[Token]) -> torch.Tensor`. It feels extremely hacky but we couldn't really figure out any other place to put this in.

In it's current form the Vocabulary / TokenIndexing is quite restrictive in that sense but also really hard to bypass since it's not configurable in the normal DatasetReaders. It's a bit hard to describe so I'll see if I can make a gist out of it that better illustrates the problem.

**Comment by joelgrus** [comment]
here is the first thing that occurs to me (I'm not sure if it's a good idea)

1. (in allennlp) change the lower bound on TokenType to allow for numpy arrays
https://github.com/allenai/allennlp/blob/master/allennlp/data/token_indexers/token_indexer.py#L7

2. (in your code) implement your logic as `MagnitudeIndexer(TokenIndexer[np.ndarray])`

[this next part is along the lines of mattg's comment]
3. (in allennlp) add a static variable `TokenIndexer.tensor_type = 'long'` (maybe there's a better name)
4. (in your code) set `MagnitudeIndexer.tensor_type = 'float'`
5. (in allennlp) modify `TextField.as_tensor` to check `token_indexer.tensor_type` and call the appropriate function.

this is a relatively small change to the library and seems to meet your need?

I'm not 100% sure how I feel about this.

**Comment by joelgrus** [comment]
this is not the first time we've wanted to "un-decouple" token indexing and token embedding, I'll also try to think more about a more general way to handle this.

**Comment by RXminuS** [comment]
On a sidenote; we're working on a PR for the Tokenizer class as well that uses `before` and `after` hooks to handle things like lowercasing, filtering, splitting etc. rather than explicit parameters for every type of Tokenizer. Maybe we could do something similar on a TextField to allow users to configure what stages to run as part of the Token -> Tensor conversion?

**Comment by RXminuS** [comment]
Actually @joelgrus you could just skip step 4 because you can get the `numpy.ndarray.dtype` directly right?

**Comment by joelgrus** [comment]
"explicit is better than implicit"

also

```
In [8]: x = np.ones(5, dtype=np.int64)                                                                                        

In [9]: t = torch.Tensor(x)                                                                                                   

In [10]: t                                                                                                                    
Out[10]: tensor([1., 1., 1., 1., 1.])

In [11]: t.dtype                                                                                                              
Out[11]: torch.float32
```

**Comment by RXminuS** [comment]
Ah, but not if you do this right? ...
```python
>>> import torch
>>> import numpy as np
>>> x = np.ones(5, dtype=np.int64)
>>> t = torch.from_numpy(x)
>>> t
tensor([1, 1, 1, 1, 1])
>>> t.dtype
torch.int64
```

**Comment by matt-gardner** [comment]
What about making a `ListField[ArrayField]` that just has your vectors already, and concatenating them to whatever other embeddings you have?  The trouble is that this doesn't fit nicely into our concept of "just use a `TextFieldEmbedder` and decide how you want to represent text later"...

I don't really see anything wrong with the solution you proposed, just looking at the numpy array dtype and selecting a tensor dtype based on it.  Then there's no need for a class variable or anything.  PR welcome.  And then you'd probably want to pair this indexer with a no-op embedder.  I thought we had one of those, but I think we only have one for encoders, not for embedders - feel free to add one of those, too.

**Comment by RXminuS** [comment]
Perfect! I think that sounds like a good workaround for now. I'm just trying it out now, so I'll push a PR in a bit (or maybe tomorrow...it's 2am here in Sweden :-) )

**Comment by RXminuS** [comment]
Ok, one thing I'm running into now though is that the TokenIndexer wasn't really designed with "heavy duty" tasks in mind. So it calls `index` per item first and then `pads` the resulting list. However, for most libraries like Magnitude or Gensim it's much more efficient to just query a whole batch of tokens in one go and pad them at the same time. At the moment I'm getting 1/10th the performance.

The tokenizer already has a similar pattern where it allows the Tokenizer instance to handle batch_tokenize. Could we introduce something similar in the TokenIndexer?

**Comment by matt-gardner** [comment]
Hmm, I'm having a hard time thinking about how to do this more efficiently, and how it would practically speed up your use case.  Any suggestions?  @joelgrus?

**Comment by joelgrus** [comment]
as mattg suggested earlier, your best bet might to do this all in the `DatasetReader` and just give each instance a `ListField[ArrayField]` representing the vectors. Inside `DatasetReader._read` you can batch however you want and bulk generate tensors and so on.

you can still use config files to switch back and forth between generating word vectors this way and generating them using the standard indexer-embedder way, it's just slightly more oblique.

this solution also doesn't require any changes to the library.

--

I think that `batch_tokenize` is not a great analogy, because it's really up to each dataset reader to explicitly know when it's appropriate for that type of instances. if anything, I think it suggests the `ListField[ArrayField]` approach.

**Comment by matt-gardner** [comment]
This still doesn't handle the padding issue - I don't think we have a good way of skipping padding if you know that all of your tensors are already the right shape.  But I'm not sure whether that's the bottleneck here.

**Comment by RXminuS** [comment]
It doesn't really allow to reuse existing DatasetReaders though? Also, it doesn't really feel natural to have embedding as part of your reader...it feels more part of your model configuration?

**Comment by RXminuS** [comment]
Having thought about it some more, that is probably the way to go then. It still doesn't feel natural to have your embeddings as part of your dataset reader (especially since that makes it impossible to re-use existing dataset readers) but can't think of any better solution for now.

**Comment by joelgrus** [comment]
on some level the challenge is that what you want isn't quite a `TextField`, which is really designed to do text -> ids -> vector; you have your own kind of field which is (text + other stuff) -> vector.

in the case you've described all the other stuff *happens* to be encapsulated into a spacy `Token` but conceptually there's no reason that has to be the case for what you're doing.

so on some level what you want is a `UnindexedTextFieldWithExtras` that takes in `Tokens` and potentially some extras (maybe `List[Any]`, I'm not sure), and whose `as_tensor` just leaves them as is (cf `MetadataField`) and then inside your model you can do whatever you want with those.

(it's possible you could just use a metadata field for this; it's also possible the padding wouldn't work right.)

**Comment by RXminuS** [comment]
Yeah, I think that's a good way of looking at it. I think that my struggle with it is that TextField sounds like it should be generically applicable; where you can easily hook in different ways of going from `text -> processing A -> processing B -> vector`. But really the current TextField is an IndexedTokenField. 

Another thing though is that a TextField is the wrong point to do a lot of the processing to go from Text -> Vector since it's hard to batch operations at that point; which is why the tokenization happens outside of it; so I think there were just several things that the abstraction layer isn't quite capturing yet

In my mind an ideal API would allow me to extract raw text from a data source using a Reader and put those into different "Text" fields, and then specify different ways of turning those fields into vector "views" for a whole batch (tokenize -> stem -> one-hot-encoding, or tokenize -> embedding lookup etc). This would allow the tuning of parameters for text -> tokens just as easy as other model parameters.

**Comment by matt-gardner** [comment]
Yeah, there are a lot of different ways to do this.  One important issue: if you want a learnable embedding, pytorch requires it to be in the _model_, so that parameters are tracked and updated correctly.  So that disallows having an embedding lookup in the data processing code.  We also wanted a clean separation between "data preparation" (that produces input tensors), and "model operations" (that have parameters), so that you can decide to only use our data code, or only use our model code, if that's what you want.  This (among many other considerations) puts some constraints on the API design.  If you have suggestions for how to improve things within the constraints that we have, we're happy to hear them.

**Comment by matt-peters** [comment]
Drive by comment to say that we have also found the split between token indexing and embedding in allennlp now to be suboptimal, and would love to see the ability to batch_index.  Our use case requires running some models as preprocessing for indexing that would be sped up considerably by allowing batch indexing.

**Comment by matt-gardner** [comment]
@matt-peters, thanks for speaking up.  We should chat about this.

**Comment by joelgrus** [comment]
batch indexing itself I think would not be that hard to implement?

because we're basically calling `Batch.index_instances(vocab)`, which is basically doing 

```
for instance in self.instances:
    instance.index_fields(vocab)
```

presumably we'd just need to "transpose" that to do "for each field, index the whole batch-worth of that field" with the default behavior just being to index one instance at a time but allowing specific fields to be more clever about it? 

I'm sure there's some edge-cases / details, but this doesn't seem like that big a change.

**Comment by matt-gardner** [comment]
I guess what I'm wondering is what bottleneck this would actually relieve.  We could pretty easily add a `as_batched_tensor` method to the `Field` API, as you say, but I'd want to be sure that this is actually solving the right problem before doing it.

**Comment by matt-peters** [comment]
Consider something that combines the traditional pipeline based NLP approach with neural nets, e.g. a model that takes tokens, POS tags, dependency parse, etc. as input features.  The POS tagger, parser and other components need to be run before indexing (to convert POS tags, or dependency labels to ids before an embedding layer).   If that POS tagger or dependency parser or whatever is itself a modern neural network it will be sped up substantially by batching.

**Comment by matt-gardner** [comment]
Yes, but predicting those things is separate from indexing and producing tensors.  Both of those are currently handled by our tokenizer, which is already batch-able.

**Comment by matt-peters** [comment]
Except many dataset readers don't use a tokenizer (ignoring the obviously misleading name of "tokenizer" for an abstraction that really does all sorts of batch preprocessing).

**Comment by matt-gardner** [comment]
You mentioned a specific use case that's slow - that's what I would like more detail about.

**Comment by matt-peters** [comment]
It's running a preprocessing process using a pre-trained neural net before indexing.  Currently the options are:

(1) preprocess every dataset off line first then create a new dataset reader to read in the original data plus the additional bits

(2) implement it as part of the Indexer (what we did), but this is very slow for the first epoch as nothing is batched.

(3) move it to the Tokenizer (which we hadn't considered), and then I suppose modify any dataset reader we want to use that doesn't use a tokenizer to add a default tokenizer that just joins then splits on whitespace so that we can hook in the custom tokenizer.

**Comment by matt-gardner** [comment]
Ok, thanks - I wasn't sure what slowness you were referring to, and I was remembering one around language modeling where it didn't seem like batching would help.

I agree that having this preprocessing in the "tokenizer" is an unfortunate naming choice, and it's only there because we have spacy as a tokenizer, and spacy can provide all of these things on the `Tokens` that it returns.
